runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5396,6 +5399,32 @@ class INTERPRETER_CLASS
 		recordJNIReturn(REGISTER_ARGS, bp);
 		restoreSpecialStackFrameLeavingArgs(REGISTER_ARGS, bp);
 
+#if JAVA_SPEC_VERSION >= 22
+		/* For each heap argument, copy the corresponding native memory
+		 * to the heap region in the case of the dicontiguous array.
+		 */
+		if (isHeapPassed) {
+			curHeapArgIdx = 0;
+			for (U_8 i = 0; i < ffiArgCount; i++) {
+				U_8 argType = LayoutFFITypeHelpers::getJ9NativeTypeCodeFromFFIType(cif->arg_types[i]);
+				if ((J9NtcPointer == argType) && (J9_FFI_DOWNCALL_HEAP_ARGUMENT_ID == (U_64)ffiArgs[i])) {
+					j9object_t heapBase = (j9object_t)J9JAVAARRAYOFOBJECT_LOAD(
+							_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 11), /* The heap base array. */
+							curHeapArgIdx);
+					if (NULL == heapBase) {
+						rc = THROW_NPE;
+						goto done;
+					}
+					if (!J9ISCONTIGUOUSARRAY(_currentThread, heapBase)) {
+						memcpyToObjectArray(_currentThread, heapBase, (void *)(pointerValues[i]), 0, JNI_FALSE);
```

Comment:
This casts directly from `U_64` to `void *`; it should be (possibly) narrowed to `UDATA` first:
```c
						memcpyToObjectArray(_currentThread, heapBase, (void *)(UDATA)(pointerValues[i]), 0, JNI_FALSE);
```

Reply:
Agreed and fixed as suggested above.